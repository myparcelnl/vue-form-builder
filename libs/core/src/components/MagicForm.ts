import {type Component, computed, defineComponent, h, onMounted, provide, ref} from 'vue';
import {type Event} from 'happy-dom';
import {get} from '@vueuse/core';
import {INJECT_FORM} from '../services';
import {FORM_HOOKS, type FormHook, type FormInstance} from '../form';
import {useLifecycleHooks} from '../composables';
import Fragment from './Fragment.vue';

//  <component
//     :is="form.config.form.tag"
//     :id="form.name"
//     v-bind="form.config.form.attributes"
//     ref="formElement"
//     @submit.prevent="() => form.submit()"
//     @reset.prevent="() => form.reset()">
//     <Fragment :component="form.config.form.wrapper">
//       <Suspense @resolve="elementsAreResolved = true">
//         <template
//           v-for="(element, index) in plainFields"
//           :key="`element--${element.name ?? 'unnamed'}--${index}`">
//           <FormElementWrapper
//             :form="form"
//             :element="element" />
//         </template>
//       </Suspense>
//
//       <template v-if="elementsAreResolved">
//         <template
//           v-for="(element, index) in teleportFields"
//           :key="`element--${element.name ?? 'unnamed'}--${index}`">
//           <FormElementWrapper
//             :form="form"
//             :element="element" />
//         </template>
//       </template>
//     </Fragment>
//   </component>

// eslint-disable-next-line @typescript-eslint/naming-convention
export const MagicForm = defineComponent({
  name: 'MagicForm',

  emits: ['submit', 'reset'],

  setup() {
    const formElement = ref<HTMLFormElement | null>(null);

    const props = defineProps<{form: FormInstance}>();

    const emit = defineEmits<(event: FormHook, form: FormInstance) => void>();

    onMounted(() => {
      const value = get(formElement);

      if (!value) {
        return;
      }

      // eslint-disable-next-line vue/no-mutating-props
      props.form.element = value;
    });

    provide(INJECT_FORM, props.form);

    FORM_HOOKS.forEach((hook) => {
      props.form.on(hook, (form) => {
        emit(hook, form);
      });
    });

    const lifeCycleHooks = useLifecycleHooks();

    lifeCycleHooks.register(props.form.hooks, props.form);

    const elementsAreResolved = ref(false);

    const plainFields = computed(() => {
      return get(props.form.fields).filter((element) => !element.teleportSelector);
    });

    const teleportFields = computed(() => {
      return get(props.form.fields).filter((element) => Boolean(element.teleportSelector));
    });

    return {
      form: props.form,
      elementsAreResolved,
      formElement,
      plainFields,
      teleportFields,
    };
  },

  render() {
    const formWrapperChildren: Component[] = [];

    return h(
      this.form.config.form.tag,
      {
        id: this.form.name,
        ...this.form.config.form.attributes,
        ref: 'formElement',

        onSubmit: (event: Event) => {
          event.preventDefault();
          void this.form.submit();
        },
        onReset: (event: Event) => {
          event.preventDefault();
          void this.form.reset();
        },
      },
      () => [
        h(
          Fragment,
          {
            component: this.form.config.form.wrapper,
          },
          () => formWrapperChildren,
        ),
      ],
    );

    //   return h(
    //     this.form.config.form.tag,
    //     {
    //       id: this.form.name,
    //     },
    //     {
    //       default: () => {
    //         return h(
    //           Fragment,
    //           {
    //             component: this.form.config.form.wrapper,
    //           },
    //           {
    //             default: () => {
    //               return h(
    //                 'template',
    //                 {
    //                   ref: 'formElement',
    //                   onSubmit: (event) => {
    //                     event.preventDefault();
    //
    //                     this.form.submit();
    //                   },
    //                   onReset: (event) => {
    //                     event.preventDefault();
    //
    //                     this.form.reset();
    //                   },
    //                 },
    //                 {
    //                   default: () => {
    //                     return h(
    //                       'suspense',
    //                       {
    //                         onResolve: () => {
    //                           this.elementsAreResolved = true;
    //                         },
    //                       },
    //                       {
    //                         default: () => {
    //                           return h(
    //                             'template',
    //                             {
    //                               v-
    //                               for: `(element, index) in plainFields`,
    //                               key: (element, index) => {
    //                                 return `element--${element.name ?? 'unnamed'}--${index}`;
    //                               },
    //                             },
    //                             {
    //                               default: () => {
    //                                 return h(
    //                                   FormElement,
    //                                   {
    //                                     form: this.form,
    //                                     element,
    //                                   },
    //                                 );
    //                               },
    //                             },
    //                           );
    //                         },
    //                       },
    //                     );
    //                   },
    //                 },
    //               );
    //             }
    //           },
    //         );
    //       },
    //
    //
    // },
  },
});
