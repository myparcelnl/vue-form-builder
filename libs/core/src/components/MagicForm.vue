<template>
  <component
    :is="form.config.form.tag"
    :id="form.name"
    v-bind="form.config.form.attributes"
    :class="{
      valid: form.isValid,
      invalid: !form.isValid,
    }"
    @submit.prevent="handleSubmit">
    <Fragment :component="form.config.form.wrapper">
      <Suspense @resolve="elementsAreResolved = true">
        <template
          v-for="(element, index) in plainFields"
          :key="`element--${element.name ?? 'unnamed'}--${index}`">
          <FormElement
            :form="form"
            :element="element" />
        </template>
      </Suspense>

      <template v-if="elementsAreResolved">
        <template
          v-for="(element, index) in teleportFields"
          :key="`element--${element.name ?? 'unnamed'}--${index}`">
          <FormElement
            :form="form"
            :element="element" />
        </template>
      </template>
    </Fragment>
  </component>
</template>

<script lang="ts">
import {PropType, computed, defineComponent, provide, ref} from 'vue';
import {AnyElementInstance} from '../types';
import FormElement from './FormElement';
import {FormInstance} from '../form';
import Fragment from './Fragment.vue';
import {INJECT_FORM} from '../services';
import {useLifeCycleHooks} from '../composables';

export default defineComponent({
  name: 'MagicForm',
  components: {
    Fragment,
    FormElement,
  },

  props: {
    form: {
      type: Object as PropType<FormInstance>,
      required: true,
    },
  },

  setup: (props) => {
    provide(INJECT_FORM, props.form);

    const elementsAreResolved = ref(false);
    const lifeCycleHooks = useLifeCycleHooks();

    lifeCycleHooks.register(props.form.hooks, props.form);

    const fields = computed(() => {
      return (props.form.fields.value ?? props.form.fields) as unknown as AnyElementInstance[];
    });

    return {
      fields,
      elementsAreResolved,
      // todo: (elements.value ?? elements) is a hack to make the tests and the actual code work.
      plainFields: computed(() => fields.value.filter((element) => !element.teleportSelector)),
      teleportFields: computed(() => fields.value.filter((element) => Boolean(element.teleportSelector))),

      async handleSubmit() {
        await props.form.submit();
      },
    };
  },
});
</script>
