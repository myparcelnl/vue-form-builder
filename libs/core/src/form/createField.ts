/* eslint-disable @typescript-eslint/naming-convention */
import {computed, defineComponent, h, markRaw, onMounted, reactive, type Ref, ref, type Component} from 'vue';
import {
  type AnyElementConfiguration,
  type AnyElementInstance,
  type ElementName,
  type ResolvedElementConfiguration,
} from '../types';
import {useForm} from '../composables';
import FormElementWrapper from '../components/FormElementWrapper';
import {generateFieldName} from './generateFieldName';

export interface CreatedField<C extends Component = Component, N extends ElementName = ElementName, RT = unknown> {
  Component: C;
  field: ResolvedElementConfiguration<C, N, RT>;
  ref: RT extends undefined ? undefined : Ref<RT>;
}

export interface ModularCreatedField<C extends Component = Component, N extends ElementName = ElementName, RT = unknown>
  extends CreatedField<C, N, RT> {
  /**
   * Only available when `wrapper` is `false`.
   * @TODO: reflect this in the type
   */
  Errors: ReturnType<typeof createErrorComponent>;

  /**
   * Only available when `wrapper` is `false`.
   * @TODO: reflect this in the type
   */
  Label: ReturnType<typeof createLabelComponent>;
}

const createMainComponent = (field: AnyElementConfiguration) => {
  return defineComponent({
    setup() {
      const form = useForm();
      const element = ref<null | AnyElementInstance>(null);

      onMounted(async () => {
        // @ts-expect-error todo
        element.value = await form.addElement(field);
      });

      return {
        element,
        form,
      };
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element);
      }

      return (
        this.element &&
        h(
          // TODO: figure out why this causes an error in the build
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          FormElementWrapper,
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          {...this.$attrs, ...this.element.props, form: this.form, element: this.element},
          this.$slots,
        )
      );
    },
  });
};

const createLabelComponent = (field: AnyElementConfiguration) =>
  defineComponent({
    // props: SHARED_PROPS,

    setup() {
      const form = useForm();
      const element = computed(() => field.name && form.getField(field.name));

      return {element};
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element, 'label');
      }

      return this.element && h('label', {for: this.element.attributes.id ?? this.element.name}, [this.element.label]);
    },
  });

const createErrorComponent = (field: AnyElementConfiguration) => {
  return defineComponent({
    setup() {
      const form = useForm();
      const element = computed(() => field.name && form.getField(field.name));

      return {element};
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element, 'errors');
      }

      return this.element?.errors && this.$slots.default?.({errors: this.element.errors});
    },
  });
};

export const createField = <C extends Component = Component, N extends ElementName = ElementName, RT = unknown>(
  field: AnyElementConfiguration<C, N, RT>,
): ModularCreatedField<C, N, RT> => {
  // @ts-expect-error todo
  return reactive({
    field,
    ref: (field.ref ?? ref<RT>()) as RT extends undefined ? undefined : Ref<RT>,
    Component: markRaw(createMainComponent(field)),

    ...(field.wrapper === false && {
      Label: markRaw(createLabelComponent(field)),
      Errors: markRaw(createErrorComponent(field)),
    }),
  });
};
