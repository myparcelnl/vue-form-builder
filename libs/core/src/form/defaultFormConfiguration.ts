import {InstanceFormConfiguration} from './Form.types';

export const DEFAULT_FORM_CONFIGURATION: InstanceFormConfiguration = {
  field: {
    elementProp: true,
  },

  fieldDefaults: {
    attributes: {},
    wrapper: true,
  },

  fields: [],

  form: {
    attributes: {},
    tag: 'form',
  },
};
