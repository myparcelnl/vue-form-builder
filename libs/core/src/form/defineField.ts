import {type Component, defineComponent, h, onMounted, ref} from 'vue';
import {
  type AnyElementConfiguration,
  type AnyElementInstance,
  type ComponentOrHtmlElement,
  type ElementName,
  type ResolvedElementConfiguration,
} from '../types';
import {useForm} from '../composables';
import Fragment from '../components/Fragment.vue';
import FormElement from '../components/FormElement.vue';

/**
 * Define a field for use in a form.
 *
 * @example defineField({name: 'name', component: 'input', props: {type: 'text'}})
 */
export const defineField = <
  C extends ComponentOrHtmlElement = ComponentOrHtmlElement,
  N extends ElementName = ElementName,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  RT = any,
>(
  field: AnyElementConfiguration<C, N, RT>,
): ResolvedElementConfiguration<C, N, RT> => {
  return field as ResolvedElementConfiguration<C, N, RT>;
};

export const defineFieldNew = <
  C extends ComponentOrHtmlElement = ComponentOrHtmlElement,
  N extends ElementName = ElementName,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  RT = any,
>(
  field: AnyElementConfiguration<C, N, RT>,
): {
  field: ResolvedElementConfiguration<C, N, RT>;
  Component: Component;
} => {
  const component = defineComponent({
    setup() {
      console.log('field', field.name, field);
      const form = useForm();
      const element = ref<null | AnyElementInstance>(null);

      onMounted(async () => {
        // @ts-expect-error todo
        element.value = await form.addElement(field);

        console.log('element', element.value);
      });

      return {
        form,
        element,
        field,
      };
    },

    render() {
      this.$options.name = `${this.form.name}:${field.name}`;

      return h(Fragment, () => [
        this.element
          ? h(
              FormElement,
              {
                form: this.form,
                element: this.element,
                field: this.field,
              },
              this.$slots,
            )
          : null,
      ]);
    },
  });

  return {
    field: field as ResolvedElementConfiguration<C, N, RT>,
    Component: component,
  };
};
