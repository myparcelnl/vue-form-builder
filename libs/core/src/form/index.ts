export * from './Form';
export * from './Form.types';
export * from './defaultFormConfiguration';
export * from './defineField';
export * from './defineForm';
export * from './interactive-element';
export * from './plain-element';
export * from './validator';
