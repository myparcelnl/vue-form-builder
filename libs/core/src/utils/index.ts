export * from './getValue';
export * from './markComponentAsRaw';
export * from './useDynamicWatcher';
