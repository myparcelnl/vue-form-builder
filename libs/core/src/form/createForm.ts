import {type Component, h, markRaw} from 'vue';
import MagicForm from '../components/MagicForm.vue';
import {defineForm} from './defineForm';
import {type FormConfiguration, type FormInstance, type FormValues} from './Form.types';

export interface CreatedForm<V extends FormValues> {
  Component: Component;
  instance: FormInstance<V>;
}

export const createForm = <V extends FormValues = FormValues>(
  name: string,
  config: Omit<FormConfiguration<V>, 'fields'>,
): CreatedForm<V> => {
  const form = defineForm(name, config as FormConfiguration<V>);

  return {
    Component: markRaw((_, ctx) => h<{form: FormInstance<V>}>(MagicForm, {form}, ctx.slots)),
    instance: form,
  };
};
