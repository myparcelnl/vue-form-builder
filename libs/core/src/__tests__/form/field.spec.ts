import {h, markRaw, ref} from 'vue';
import {describe, expect, it, vi} from 'vitest';
import {flushPromises, mount} from '@vue/test-utils';
import {generateForm, optionData} from '../utils';
import {firstNameNotDuane} from '../examples/validators';
import TextInput from '../elements/TextInput.vue';
import FormGroup from '../elements/FormGroup.vue';
import {defineField, defineForm} from '../../form';
import MagicForm from '../../components/MagicForm.vue';

describe('Form fields', () => {
  it.skip('can calculate forwards based on primary input', async () => {
    const firstName = ref('');
    const price = ref('0');

    const validationForm = defineForm('validationForm', {
      field: {
        wrapper: markRaw(FormGroup),
      },
      fields: [
        defineField({
          name: 'firstName',
          component: TextInput,
          ref: firstName,
          afterUpdate: (field, newValue) => {
            field.form.model.price.ref.value = newValue === 'Jack' ? '100' : '50';
          },
        }),
        defineField({
          name: 'price',
          component: TextInput,
          ref: price,
        }),
      ],
    });

    const wrapper = mount(MagicForm, {props: {form: validationForm}});
    await flushPromises();

    await wrapper.find('input[name="firstName"]').setValue('Hank');
    expect(price.value).toBe('50');

    await wrapper.find('input[name="firstName"]').setValue('Jack');
    expect(price.value).toBe('100');
  });

  it('can calculate forwards based on primary input, out of a promise', async () => {
    expect.assertions(2);
    vi.useFakeTimers();

    const firstName = ref('');
    const price = ref('0');

    const validationForm = defineForm('validationForm', {
      fields: [
        defineField({
          name: 'firstName',
          component: TextInput,
          ref: firstName,
          afterUpdate: async (field, newValue) => {
            const remote = await optionData(newValue);

            field.form.model.price.ref.value = remote.price;
          },
        }),
        defineField({
          name: 'price',
          component: TextInput,
          ref: price,
        }),
      ],
    });

    const wrapper = mount(MagicForm, {props: {form: validationForm}});
    await flushPromises();
    await wrapper.find('input[name="firstName"]').setValue('Hank');
    vi.advanceTimersByTime(1000);

    await flushPromises();
    expect(price.value).toBe('50');
    await wrapper.find('input[name="firstName"]').setValue('John');

    vi.advanceTimersByTime(1000);
    await flushPromises();
    expect(price.value).toBe('100');
    vi.useRealTimers();
  });

  it.skip('can calculate backwards based on primary input, out of a promise', async () => {
    vi.useFakeTimers();
    const firstName = ref('');
    const price = ref('0');

    const validationForm = defineForm('validationForm', {
      fields: [
        defineField({
          name: 'firstName',
          component: TextInput,
          ref: firstName,
        }),
        defineField({
          name: 'price',
          component: TextInput,
          ref: price,
        }),
      ],
    });
    const wrapper = mount(MagicForm, {props: {form: validationForm}});
    await flushPromises();

    await wrapper.find('input[name="firstName"]').setValue('Hank');
    // figure out how to reverse calculate the price when firstName is updated.
  });

  it.skip('can teleport a field in the same form', async () => {
    // expect.assertions(2);
    const form = generateForm({
      fields: [
        defineField({
          name: 'name',
          component: h('div', {class: 'flex flex-row gap-2'}, [
            h('div', {id: 'teleport--firstname'}),
            h('div', {id: 'teleport--lastname'}),
          ]),
          label: 'name',
          props: {
            label: 'Naam',
          },
        }),
        defineField({
          name: 'firstname',
          component: 'input',
          ref: ref(''),
          label: 'firstname',
          teleportSelector: '#teleport--firstname',
        }),

        defineField({
          name: 'lastname',
          component: 'input',
          ref: ref(''),
          label: 'lastname',
          teleportSelector: '#teleport--lastname',
        }),
      ],
    });
    const wrapper = mount(MagicForm, {
      global: {
        stubs: {
          teleport: true,
        },
      },
      props: {
        form,
      },
    });
    await flushPromises();
  });

  it('can be reset', async () => {
    expect.assertions(1);

    const field = defineField({
      component: 'input',
      name: 'element',
      ref: ref(''),
      validators: [firstNameNotDuane()],
    });

    const form = generateForm({fields: [field]});

    await form.model.element.reset();

    expect(form.isValid.value).toBe(true);
  });

  it('gets filled with initial data from form config', async () => {
    expect.assertions(1);

    const field = defineField({
      component: 'input',
      name: 'element',
      ref: ref('disregarded'),
    });

    const form = generateForm({fields: [field], initialValues: {element: 'hello'}});

    expect(form.model.element.ref.value).toBe('hello');
  });
});
