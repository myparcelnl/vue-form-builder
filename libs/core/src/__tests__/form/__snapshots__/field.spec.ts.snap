// Vitest Snapshot v1

exports[`Form fields > can calculate backwards based on primary input, out of a promise 1`] = `
"<form id=\\"validationForm\\" class=\\"valid\\"><input id=\\"firstName\\" type=\\"text\\" name=\\"firstName\\" class=\\"\\">
  <!--v-if-->
  <!--v-if--><input id=\\"price\\" type=\\"text\\" name=\\"price\\" class=\\"\\">
  <!--v-if-->
  <!--v-if-->
</form>"
`;

exports[`Form fields > can calculate forwards based on primary input 1`] = `
"<form id=\\"validationForm\\" class=\\"valid\\"><input id=\\"firstName\\" type=\\"text\\" name=\\"firstName\\" class=\\"\\">
  <!--v-if-->
  <!--v-if--><input id=\\"price\\" type=\\"text\\" name=\\"price\\" class=\\"\\">
  <!--v-if-->
  <!--v-if-->
</form>"
`;
